backends-velox/src/main/scala/org/apache/spark/sql/execution/BroadcastUtils.scala

```diff
@@ -0,0 +1,120 @@
+
+import org.apache.spark.SparkContext
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.plans.physical.{BroadcastMode, BroadcastPartitioning, Partitioning}
+import org.apache.spark.sql.execution.joins.HashedRelation
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.util.TaskResources
+
+import scala.collection.JavaConverters._;
+
+// Utility methods to convert Vanilla broadcast relations from/to Velox broadcast relations.
+// FIXME: Add checking for broadcast mode.
+// FIXME: Truncate output with batch size.
+object BroadcastUtils {
+  def veloxToSparkUnsafe[F, T](
+      context: SparkContext,
+      mode: BroadcastMode,
+      from: Broadcast[F],
+      fn: Iterator[ColumnarBatch] => Iterator[InternalRow]): Broadcast[T] = {
+    // ColumnarBuildSideRelation to HashedRelation.
+    val fromBroadcast = from.asInstanceOf[Broadcast[ColumnarBuildSideRelation]]
+    val fromRelation = fromBroadcast.value.asReadOnlyCopy()
+    var rowCount: Long = 0
+    val toRelation = TaskResources.runUnsafe {
+      val rowIterator = fn(fromRelation.deserialized.flatMap {
+        cb =>
+          rowCount += cb.numRows()
+          Iterator(cb)
+      })
+      mode.transform(rowIterator, Some(rowCount))
+    }
+    // Rebroadcast Spark relation.
+    context.broadcast(toRelation).asInstanceOf[Broadcast[T]]
+  }
+
+  def sparkToVeloxUnsafe[F, T](
+      context: SparkContext,
+      mode: BroadcastMode,
+      schema: StructType,
+      from: Broadcast[F],
+      fn: Iterator[InternalRow] => Iterator[ColumnarBatch]): Broadcast[T] = {
+    // HashedRelation to ColumnarBuildSideRelation.
+    val fromBroadcast = from.asInstanceOf[Broadcast[HashedRelation]]
+    val fromRelation = fromBroadcast.value.asReadOnlyCopy()
+    val keys = fromRelation.keys()
+    val toRelation = TaskResources.runUnsafe {
+      val batchItr: Iterator[ColumnarBatch] = fn(keys.flatMap(key => fromRelation.get(key)))
+      val serialized: Array[Array[Byte]] = serializeStream(batchItr) match {
+        case ColumnarBatchSerializeResult.EMPTY =>
+          Array()
+        case result: ColumnarBatchSerializeResult =>
+          Array(result.getSerialized)
+      }
+      ColumnarBuildSideRelation(schema.toAttributes, serialized)
+    }
+    // Rebroadcast Velox relation.
+    context.broadcast(toRelation).asInstanceOf[Broadcast[T]]
+  }
+
+  def getBroadcastMode(partitioning: Partitioning): BroadcastMode = {
+    partitioning match {
+      case BroadcastPartitioning(mode) =>
+        mode
+      case _ =>
+        throw new IllegalArgumentException("Unexpected partitioning: " + partitioning.toString)
+    }
+  }
+
+  def serializeStream(batches: Iterator[ColumnarBatch]): ColumnarBatchSerializeResult = {
+    val filtered = batches
+      .filter(_.numRows() != 0)
+      .map(
+        b => {
+          ColumnarBatches.retain(b)
```

Comment:
Can there be out of memory scenarios? since the original broadcast will be  there in heap memory and then again memory will be allotted until the serialization is finished.

Reply:
If I understand correctly, you are considering the amplified memory consumption by the original hash relation + converted columnar batch + serialized byte arrays?

If so, it should become a problem when the broadcast threshold is set to a large number. We can optimize it in further development iterations with some benchmarks.